changefeedccl: remove duplicate columns from parquet output

aerfrei opened the pull request:
Previously, in changefeeds using CDC queries and parquet,
we would see duplicate columns in the output when using a
user defined primary key. This was confusing and unexpected.
This change has us deduplicating columns in the output when
writing to parquet.

Epic: none

Fixes: https://github.com/cockroachdb/cockroach/issues/124434

Release note (bug fix): Removes duplicate columns in the parquet
output from changefeeds using cdc queries.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/136718)
<!-- Reviewable:end -->

aerfrei commented:
**Behaviour:** 
This code change has us outputting the columns from a cdcevent row with ForEachColumn instead of ForAllColumns. The difference between these is that ForEachColumn iterates over just the “valueCols” and not “allCols” which include the key columns. This is the set of columns iterated over for both JSON and CSV outputs. For our purposes, this ensures that in the example in #124434 we no longer output the primary key column twice (once as a primary key column and once as a value column).

Note that this means that the output to parquet may not include the primary key of the table if not specified by the user as in this example:
```
`CREATE TABLE t (id INT8)`
`INSERT INTO t VALUES (9)`
`CREATE CHANGEFEED WITH format=parquet,initial_scan='only' AS SELECT id FROM t`
```
The “rowId” column which is the primary key will not be included in the parquet output.

**Testing:** 
Unfortunately, this issue is hidden by the assertPayloads test function because that converts the changefeed message to JSON, which doesn’t allow for duplicate keys. 

It’s useful that the format of these messages is in JSON since tests like pkg/ccl/changefeedccl:changefeedccl_test seem to be able to run against both parquet and non-parquet changefeeds. 

Instead, we can simply assert that the number of distinct columns in the parquet test metadata matches the number of datums we see in the row when writing a test message. If there were a duplicate, we do not expect those numbers to match so the test will fail with an error. This allows us to catch our edge case here while keeping the test feed message format uniform.

Even in cases like this one:
```
`CREATE CHANGEFEED WITH format=parquet,initial_scan='only' AS SELECT id, id FROM t`
```
where a column is repeated in the output, we expect those columns to be given unique names in the metadata.

rharding6373 reviewed (commented):
Nice find!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136718#-:-ODJ3NFo4Yt6q0jChaAa:bdkgtm2)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aerfrei)

___
*[`pkg/ccl/changefeedccl/testfeed_test.go` line 1327 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136718#-ODJ112s5W29tE-h7u8e:-ODJ112s5W29tE-h7u8f:bm56fsf) ([raw file](https://github.com/cockroachdb/cockroach/blob/18845e2214413fb3dd958d5efd60da8c565b2bfe/pkg/ccl/changefeedccl/testfeed_test.go#L1327)):*
> ```Go
> 		if len(row) != len(valueColumnNamesOrdered) {
> 			// Throw this error if datums are duplicated in the parquet output
> 			return errors.New("More datums than columns in parquet row")
> ```

There could be fewer datums, too, right? Also, it would be useful to include information like "expected X columns, got Y columns" in this error.
___
*[`pkg/ccl/changefeedccl/testfeed_test.go` line 1336 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136718#-ODJ2ImM0CxP_E65ZlpH:-ODJ2ImM0CxP_E65ZlpI:b-5t7u9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/18845e2214413fb3dd958d5efd60da8c565b2bfe/pkg/ccl/changefeedccl/testfeed_test.go#L1336)):*
> ```Go
> 				// a CDC query, it is expected that the output row doesn't contain the
> 				// primary key.
> 				continue
> ```

Based on the test failures, it looks like this approach may not work with column families. Or maybe some existing tests need to be updated to not compare primary keys with column families.


<!-- Sent from Reviewable.io -->

aerfrei committed:
changefeedccl: remove duplicate columns from parquet output

Previously, in changefeeds using CDC queries and parquet,
we would see duplicate columns in the output when using a
user defined primary key. This was confusing and unexpected.
This change has us deduplicating columns in the output when
writing to parquet.

Epic: none

Fixes: #124434

Release note (bug fix): Removes duplicate columns in the parquet
output from changefeeds using cdc queries.

aerfrei commented:
> There could be fewer datums, too, right? Also, it would be useful to include information like "expected X columns, got Y columns" in this error.
@rharding6373 Good point, thanks! Updated this.
Also changed the way this was working because of the column families issue.

andyyang890 reviewed (commented):
I'm going to defer to Miles for final approval since he's the parquet expert.

also process nit: if there are reviewable comments, you should continue commenting in reviewable

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136718#-:-ODhXqJy4gzIud-EKOTJ:b-spmd6n)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aerfrei, @asg0451, and @wenyihu6)

___
*[`-- commits` line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136718#-ODhW7OD7dk60XLzbV8R:-ODhW7OD7dk60XLzbV8S:b-15r9b8):*
nit: generally you should use the most specific package name that's still broad enough to capture the majority of the changes, i.e. `changefeedccl` here
___
*[`pkg/ccl/changefeedccl/parquet_test.go` line 334 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136718#-ODhc83rBg95YLsRNGfZ:-ODhc83rBg95YLsRNGf_:b20pwm2) ([raw file](https://github.com/cockroachdb/cockroach/blob/865f051c843245771283a8f213537224a82a3221/pkg/ccl/changefeedccl/parquet_test.go#L334)):*
> ```Go
> }
> 
> func TestParquetSpecifiedDuplicateQueryColumns(t *testing.T) {
> ```

Could we add a similar test for duplicate non-primary key columns?
___
*[`pkg/ccl/changefeedccl/parquet_test.go` line 349 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136718#-ODhbg8C-uCmcxhth0Q6:-ODhbg8C-uCmcxhth0Q7:b-s3h03l) ([raw file](https://github.com/cockroachdb/cockroach/blob/865f051c843245771283a8f213537224a82a3221/pkg/ccl/changefeedccl/parquet_test.go#L349)):*
> ```Go
> 		// columns, Expected: %d, Recieved: %d`.
> 		assertPayloads(t, foo, []string{
> 			`t: [1]->{"id": 1, "id_1": 1}`,
> ```

Out of curiosity, where in the code is the suffixing logic for duplicate columns?


<!-- Sent from Reviewable.io -->

aerfrei reviewed (commented):
Thanks for the tip!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136718#-:-ODl8A80B8GtbyCI4haw:b-p1bteq)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @asg0451, @rharding6373, and @wenyihu6)

___
*[`pkg/ccl/changefeedccl/parquet_test.go` line 334 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136718#-ODhc83rBg95YLsRNGfZ:-ODl9ksI-HdNkHFg1MFy:bfnxy98) ([raw file](https://github.com/cockroachdb/cockroach/blob/865f051c843245771283a8f213537224a82a3221/pkg/ccl/changefeedccl/parquet_test.go#L334)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

Could we add a similar test for duplicate non-primary key columns?
</blockquote></details>

I added this into this test if that's ok
___
*[`pkg/ccl/changefeedccl/parquet_test.go` line 349 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136718#-ODhbg8C-uCmcxhth0Q6:-ODl7wrv67A5enDIxGVz:bo9r10o) ([raw file](https://github.com/cockroachdb/cockroach/blob/865f051c843245771283a8f213537224a82a3221/pkg/ccl/changefeedccl/parquet_test.go#L349)):*
<details><summary><i>Previously, andyyang890 (Andy Yang) wrote…</i></summary><blockquote>

Out of curiosity, where in the code is the suffixing logic for duplicate columns?
</blockquote></details>

It's in setupProjection in pkg/ccl/changefeedccl/cdceval/expr_eval.go
___
*[`pkg/ccl/changefeedccl/testfeed_test.go` line 1327 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136718#-ODJ112s5W29tE-h7u8e:-ODl7hVUEPpXcI72ZYaM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/18845e2214413fb3dd958d5efd60da8c565b2bfe/pkg/ccl/changefeedccl/testfeed_test.go#L1327)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

There could be fewer datums, too, right? Also, it would be useful to include information like "expected X columns, got Y columns" in this error.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/testfeed_test.go` line 1336 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136718#-ODJ2ImM0CxP_E65ZlpH:-ODl7gbD0wNtaTT57QCU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/18845e2214413fb3dd958d5efd60da8c565b2bfe/pkg/ccl/changefeedccl/testfeed_test.go#L1336)):*
<details><summary><i>Previously, rharding6373 (Rachael Harding) wrote…</i></summary><blockquote>

Based on the test failures, it looks like this approach may not work with column families. Or maybe some existing tests need to be updated to not compare primary keys with column families.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

asg0451 commented on pkg/ccl/changefeedccl/testfeed_test.go:
```diff
@@ -1318,10 +1318,42 @@ func (c *cloudFeed) appendParquetTestFeedMessages(
+	// that it matches the number of datums we output. Duplicating them in the
+	// output or missing one of these keys in the parquet output would be
+	// unexpected.
+	numDistinctColumns := 0
```
nit: looks to me like you can just use len(seenColumnNames) instead of keeping track of the num yourself.

also nit: a set should be map[something]struct{}, not bool. unless you're actually using the bool which you dont appear to be doing here

aerfrei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136718)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @andyyang890, @asg0451, @rharding6373, and @wenyihu6)



<!-- Sent from Reviewable.io -->

aerfrei commented:
bors r=asg0451

andyyang890 commented:
@aerfrei You'll also need to backport this to `release-23.2`. Backporting to `release-23.2.20-rc` only includes it in that patch release. In the future, you could first backport it to the release branch first and then backport it to the RC branch from the release branch.

aerfrei commented:
blathers backport 23.2
